File: 61.rotate-list.py
#
# @lc app=leetcode id=61 lang=python3
#
# [61] Rotate List
#

# Time:  O(n)
# Space: O(1)


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# @lc code=start
class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head or not head.next:
            return head

        curr, n = head, 1
        while curr.next:
            curr, n = curr.next, n + 1
        curr.next = head

        curr, tail = head, curr
        for _ in range(n - k % n):
            tail = curr
            curr = curr.next
        tail.next = None
        return curr


# @lc code=end

File: 61.rotate-list.js
/*
 * @lc app=leetcode id=61 lang=javascript
 *
 * [61] Rotate List
 */

function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

// @lc code=start
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function (head, k) {
  if (!head || !head.next) {
    return head;
  }
  let curr = head;
  let len = 1;
  while (curr.next) {
    curr = curr.next;
    ++len;
  }
  curr.next = head;
  for (let i = 0; i < len - (k % len); ++i) {
    curr = curr.next;
  }
  const newHead = curr.next;
  curr.next = null;
  return newHead;
};
// @lc code=end
